### ISSUE TITLE
[Protocol] Implement Session Rollovers

### ISSUE THREAD

**Issue by Olshansk** [pokt-network/poktroll#275]
## Objective

Resolve the infamous session rollover issue once and for all

## Origin Document

<img width="827" alt="Screenshot 2023-12-14 at 3 54 21 PM" src="https://github.com/pokt-network/poktroll/assets/1892194/411aec2c-4091-4079-b306-8e7773af3ed7">

While working on #252, we had to add a temporary hack below due to:
- Very fast block times
- Small number of blocks per session

Overall, this is us hitting the session rollover issue in Shannon

```go
	// TODO_IN_THIS_PR: Either slow down blocks, or increase numBlocksPerSession,
	// or create a ticket related to session rollovers and link to it here.
	// currentBlock := rp.blockClient.LastNBlocks(ctx, 1)[0]
	// session, err := rp.sessionQuerier.GetSession(ctx, appAddress, service.Id, currentBlock.Height())
	// session, err := rp.sessionQuerier.GetSession(ctx, appAddress, service.Id, relayRequest.Meta.SessionHeader.SessionStartBlockHeight)
	currentBlock := rp.blockClient.LastNBlocks(ctx, 1)[0]
	session, err := rp.sessionQuerier.GetSession(ctx, appAddress, service.Id, currentBlock.Height())
```

## Goals

- Design & Implement an on-chain solution for session rollovers in Shannon

## Deliverables

- [ ] A PR that implements and solves session rollovers
- [ ] A governance parameter that modulates how much leeway session rollovers have
- [ ] Document how session rollovers are solved in Shannon

## Non-goals / Non-deliverables

- Using the same solution for session rollovers as in Morse (i.e. client side)

## General deliverables

- [ ] **Comments**: Add/update TODOs and comments alongside the source code so it is easier to follow.
- [ ] **Testing**: Add new tests (unit and/or E2E) to the test suite.
- [ ] **Makefile**: Add new targets to the Makefile to make the new functionality easier to use.
- [ ] **Documentation**: Update architectural or development READMEs; use [mermaid](https://mermaid-js.github.io/mermaid/) diagrams where appropriate.

---

**Creator**: @Olshansk 
**Co-Owners**: @red-0ne

**Comment by Olshansk** [comment]
tl;dr In `pkg/relayer/proxy/relay_verifier.go`:

```diff
+  session, err := rp.sessionQuerier.GetSession(ctx, appAddress, service.Id, relayRequest.Meta.SessionHeader.SessionStartBlockHeight)
- currentBlock := rp.blockClient.LastNBlocks(ctx, 1)[0]
- session, err := rp.sessionQuerier.GetSession(ctx, appAddress, service.Id, currentBlock.Height())
```

---

While implementing the quickstart guide in #252, the change above was necessary to make things work E2E because the session was ending too quickly.

Until session rollovers are implemented, we can potentially run into bugs / issues in local development as well. The solution is to either:
1. Add the diff above (the easier temp workaround)
2. Locally increase blocks per session (not gov param at the time of writing this)
3. Reducing block time (a rollkit flag)

**Comment by red-0ne** [comment]
Initial investigation and solution proposal https://www.notion.so/buildwithgrove/Sessions-rollover-f2845bd812c24b9c98934ca510bb70e9

**Comment by Olshansk** [comment]
> Initial investigation and solution proposal [notion.so/buildwithgrove/Sessions-rollover-f2845bd812c24b9c98934ca510bb70e9](https://www.notion.so/buildwithgrove/Sessions-rollover-f2845bd812c24b9c98934ca510bb70e9)

Reviewed!

Overall, the idea I had in mind was a little different. Only copy-pasting the mermaid diagram here for reference:


![Screenshot 2024-01-12 at 3 39 29 PM](https://github.com/pokt-network/poktroll/assets/1892194/4adfdbd9-e2cd-4a99-a971-e06fbf83e29d)



CCing @bryanchriswhite since this may be dependant on the work in https://github.com/pokt-network/poktroll/issues/141#issuecomment-1844261492

**Comment by red-0ne** [comment]
Closing as per #326 and #332
